File: app/src/main/java/com/gnoemes/shikimoriapp/domain/search/SearchInteractorImpl.java
package com.gnoemes.shikimoriapp.domain.search;

import android.support.annotation.NonNull;

import com.gnoemes.shikimoriapp.data.repository.search.SearchRepository;
import com.gnoemes.shikimoriapp.entity.anime.domain.Anime;
import com.gnoemes.shikimoriapp.entity.search.domain.FilterItem;
import com.gnoemes.shikimoriapp.utils.rx.RxUtils;
import com.gnoemes.shikimoriapp.utils.rx.SingleErrorHandler;

import java.util.HashMap;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Single;

public class SearchInteractorImpl implements SearchInteractor {

    private SearchRepository searchRepository;
    private SearchQueryBuilder queryBuilder;
    private SingleErrorHandler singleErrorHandler;
    private RxUtils rxUtils;


    @Inject
    public SearchInteractorImpl(@NonNull SearchRepository searchRepository,
                                @NonNull SearchQueryBuilder queryBuilder,
                                @NonNull SingleErrorHandler singleErrorHandler,
                                @NonNull RxUtils rxUtils) {
        this.searchRepository = searchRepository;
        this.queryBuilder = queryBuilder;
        this.singleErrorHandler = singleErrorHandler;
        this.rxUtils = rxUtils;
    }

    @Override
    public Single<List<Anime>> loadListWithFilters(HashMap<String, List<FilterItem>> filters, int page, int limit) {
        return queryBuilder.createQueryFromFilters(filters, page, limit)
                .flatMap(queryMap -> searchRepository.getAnimeList(queryMap))
                .compose((SingleErrorHandler<List<Anime>>) singleErrorHandler)
                .compose(rxUtils.applySingleSchedulers());
    }

    @Override
    public Single<List<Anime>> loadList(int page, int limit) {

        return queryBuilder.createQueryFromFilters(null, page, limit)
                .flatMap(queryMap -> searchRepository.getAnimeList(queryMap))
                .compose((SingleErrorHandler<List<Anime>>) singleErrorHandler)
                .compose(rxUtils.applySingleSchedulers());
    }
}

File: app/src/main/java/com/gnoemes/shikimoriapp/domain/search/SearchInteractorImpl.kt
package com.gnoemes.shikimoriapp.domain.search

import com.gnoemes.shikimoriapp.data.repository.search.SearchRepository
import com.gnoemes.shikimoriapp.entity.anime.domain.Anime
import com.gnoemes.shikimoriapp.entity.manga.domain.Manga
import com.gnoemes.shikimoriapp.entity.roles.domain.Character
import com.gnoemes.shikimoriapp.entity.roles.domain.Person
import com.gnoemes.shikimoriapp.entity.search.domain.FilterItem
import com.gnoemes.shikimoriapp.utils.rx.RxUtils
import com.gnoemes.shikimoriapp.utils.rx.SingleErrorHandler
import io.reactivex.Single
import javax.inject.Inject

class SearchInteractorImpl @Inject constructor(
        private val repository: SearchRepository,
        private val queryBuilder: SearchQueryBuilder,
        private val errorHandler: SingleErrorHandler<Any>,
        private val rxUtils: RxUtils
) : SearchInteractor {

    override fun loadAnimeList(page: Int, limit: Int): Single<List<Anime>> =
            queryBuilder.createQueryFromFilters(null, page, limit)
                    .flatMap { repository.getAnimeList(it) }
                    .composeHandlerAndSchedulers()

    override fun loadMangaList(page: Int, limit: Int): Single<List<Manga>> =
            queryBuilder.createQueryFromFilters(null, page, limit)
                    .flatMap { repository.getMangaList(it) }
                    .composeHandlerAndSchedulers()

    override fun loadRanobeList(page: Int, limit: Int): Single<List<Manga>> =
            queryBuilder.createQueryFromFilters(null, page, limit)
                    .flatMap { repository.getMangaList(it) }
                    .composeHandlerAndSchedulers()

    override fun loadCharacterListWithFilters(filters: HashMap<String, MutableList<FilterItem>>?, page: Int, limit: Int): Single<List<Character>> =
            queryBuilder.createQueryFromFilters(filters, page, limit)
                    .flatMap { repository.getCharacterList(it) }
                    .composeHandlerAndSchedulers()

    override fun loadPersonListWithFilters(filters: HashMap<String, MutableList<FilterItem>>?, page: Int, limit: Int): Single<List<Person>> =
            queryBuilder.createQueryFromFilters(filters, page, limit)
                    .flatMap { repository.getPersonList(it) }
                    .composeHandlerAndSchedulers()

    override fun loadAnimeListWithFilters(filters: HashMap<String, MutableList<FilterItem>>?, page: Int, limit: Int): Single<List<Anime>> =
            queryBuilder.createQueryFromFilters(filters, page, limit)
                    .flatMap { repository.getAnimeList(it) }
                    .composeHandlerAndSchedulers()

    override fun loadMangaListWithFilters(filters: HashMap<String, MutableList<FilterItem>>?, page: Int, limit: Int): Single<List<Manga>> =
            queryBuilder.createQueryFromFilters(filters, page, limit)
                    .flatMap { repository.getMangaList(it) }
                    .composeHandlerAndSchedulers()

    override fun loadRanobeListWithFilters(filters: HashMap<String, MutableList<FilterItem>>?, page: Int, limit: Int): Single<List<Manga>> =
            queryBuilder.createQueryFromFilters(filters, page, limit)
                    .flatMap { repository.getRanobeList(it) }
                    .composeHandlerAndSchedulers()


    private fun <T> Single<T>.composeHandlerAndSchedulers(): Single<T> {
        return this.compose(errorHandler as SingleErrorHandler<T>)
                .compose(rxUtils.applySingleSchedulers())
    }
}